### PR TITLE
add support for indices

### LINKED ISSUE #20737: Add support for indices
We should add another checkbox to the field `schema` setup pane in settings to mark the current field as an `index`. Knex has an existing chainable method `.index()` that works the same as `.unique()`, so I'm hopeful this is doable by effectively duplicating what we're doing for `unique`, and updating it to `index`.

### PR DESCRIPTION
**Pull request by mahendraHegde**
## Scope

What's changed:

Added support to mark a field as indexed.

- [x] add new field `is_indexed` to schema
- [x] add dialects for 
    - [x] sqlite
    - [x] pg  
    - [x] mysql
    - [x] cocroachdb
    - [x] mssql
    - [x] oracledb
 - [x] change UI to show new checkbox for the same
 - [x] add i18n changes
 - [x] add/modiy tests 

## Potential Risks / Drawbacks
NA

## Review Notes / Questions
added at relevant places as review
---

Fixes #20737

#### Add index
 request
Screenshot 2023-12-29 at 2 37 49 PM

response
Screenshot 2023-12-29 at 2 38 02 PM

#### Remove index
request
Screenshot 2023-12-29 at 2 38 34 PM

response
Screenshot 2023-12-29 at 2 38 50 PM

### CODE REVIEW

File: app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue
Comment (mahendraHegde) [inline app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue:453]: does it make sense to allow both unique and index for a column? i feel its redundant

File: packages/schema/src/dialects/postgres.ts
Comment (mahendraHegde) [inline packages/schema/src/dialects/postgres.ts:418]: idea is to select 1 non composite, non unique row from `pg_index` catalog

File: packages/schema/src/dialects/mysql.ts
Comment (mahendraHegde) [inline packages/schema/src/dialects/mysql.ts:57]: `MUL` is for non unique index(including composite), I see directus [doesnt support composite indexes](https://github.com/directus/directus/issues/19377) today as well, so i went ahead with this, another option is to use `SEQ_IN_INDEX` from `INFORMATION_SCHEMA.STATISTICS` , which is not straight forward either

File: app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue
Comment (mahendraHegde) [reply app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue:453]: in fact oracle doesnt allow both unique and normal index on a column.

File: packages/schema/src/dialects/oracledb.ts
Comment (mahendraHegde) [inline packages/schema/src/dialects/oracledb.ts:317]: `user_indexes` doesnt have filter by column, so had to use intermediate table `user_ind_columns`.

Comment (mahendraHegde) [reply packages/schema/src/dialects/oracledb.ts]: also this is not composite index safe, to make it composite index same, we'd have to partition by index_name(as its same for multiple columns), since directus doesn't support composite indexes anyway, i went ahead with this to save a tiny bit of performance.

**Comment by mahendraHegde** [comment]
@azrikahar could you review this plz, i will write i18n transalations once you are okay with the wordings, also is there a way to generate translations? or do you manually write it?

**Review by rithviknishad — APPROVED** [review]
LGTM, can we get this merged?

**Review by azrikahar — COMMENTED** [review]
Thanks for working on this! One scenario that I think we will need to address is for existing index (i.e. which are added directly to the database, not via Directus).

As knex currently creates and drops index based on the default [`<table>_<column>_index` naming convention](https://github.com/knex/knex/blob/4ca3dd5bc28e0665c5bed55026fac2ec45489d81/lib/schema/tablecompiler.js#L389-L403), currently when there's an existing index named such as `INDEX_ABC`, the UI does show the field's index is enabled. However when we try to remove it, it will return error that it was unable to drop it since it couldn't find the "default-named" index:

(example with a PostgreSQL setup)

image

The two options that I can think of at the moment would be:

- add a schema helper function to retrieve the name of the index so that we can pass it to `dropIndex()` as the second argument
- only detect index with the default naming convention in the schema package, so that we'll technically ignore index that aren't added via Directus (or technically Knex), unless we'd want to roll our own index naming convention

The first option would probably be the cleaner approach, but any thoughts on this?

> also is there a way to generate translations? or do you manually write it?

No worries, your additions to `en-US.yaml` will suffice 👍 the new translation keys will then appear via Crowdin so other community members can add their translations to other locales.

**Comment by mahendraHegde** [comment]
@azrikahar, agreed 1st approach is cleaner, 
in fact i had thought about it, since its true for many cases as directus doesnt really fully accept direct DB changes(composite indexes or rename primary index or probably even unique index), i thought this implementation is acceptable.

as a 3rd option, what do you think about changing `is_indexed` to something like `single_index:${index_name}`, bool check will look for falseness to  validate if index exist or not, and while dropping we can direct use the field as it holds index name. this way we dont need to make another db call to fetch the index.

**Comment by phazonoverload** [comment]
> LGTM, can we get this merged?

Thanks for reviewing @rithviknishad. Just want to politely note that all contributions needs review from a member of the Core Team and community reviews don’t change that requirement.

**Comment by azrikahar** [comment]
> as a 3rd option, what do you think about changing `is_indexed` to something like `single_index:${index_name}`, bool check will look for falseness to validate if index exist or not, and while dropping we can direct use the field as it holds index name. this way we dont need to make another db call to fetch the index.

That's very interesting! My second option was technically intended to prevent user from dropping manual-added index "unintentionally" (depends on the user of course), since they wouldn't know what exact index is being removed. But with your suggested 3rd option, we may be able to also include the index name in the v-checkbox label when it's checked, so that the user would be able to know the exact index that is going to be dropped 👍 This 3rd option would indeed be the better approach, sounds good to me!

**Comment by mahendraHegde** [comment]
@azrikahar i have implemented the 3rd approach share your thoughts plz.

File: app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue
Comment (ComfortablyCoding) [reply app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue]: We should not restrict unique + index for all drivers due to one driver. We should allow this for all drivers, if the driver does not support both (i.e. oracle) then if one is active and the other is selected then it should be ignored but the API should return both as active.

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]
This looks great and thanks for the contribution! This seems to work for me from a functional standpoint.

Some feedback: 
- Instead of using an object we should use a single variable named `indexed` (or equivalent). 
   - indexed as `null` means the field has no index
   - indexed with a value that is the index name
- Docs should be updated with the new property for schema.
- Tests should be added to cover the new cases.
- I believe the `simple_index_name` is somewhat redundant, what was the reasoning behind adding this for some drivers?

File: app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue
Comment (mahendraHegde) [reply app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue]: ok thats a good suggestion, thanks. i will return both values if one of them is true.

**Comment by mahendraHegde** [comment]
> Instead of using an object we should use a single variable named indexed.
indexed as null means the field has no index
indexed with a value that is the index name

hey @ComfortablyCoding  thanks for the review. I thought of using only 1 field instead object, however when any field is updated, UI sends the indexed as null, and we wont know if the intention is to remove the index or just a null, also its consistent with other fields(ex: unique), we wont remove the index unless ui sends false.

>Docs should be updated with the new property for schema.

Oh I missed, thanks for pointing, will add.

>Tests should be added to cover the new cases.

i dint see any tests for the files i have made changes to, which tests are you pointing to?

>I believe the simple_index_name is somewhat redundant, what was the reasoning behind adding this for some drivers?

i will change it to `INDEX_NAME` to keep consistent with other drivers.

File: app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue
Comment (ComfortablyCoding) [reply app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue]: If it is possible it is better if both are returned as true only if the index was selected as well.

e.g.
- unique or index selected -> all drivers return only the respective field as true
- unique + index are selected
   - driver supports both -> both returned as true and set at db level
   - driver does not support both -> both returned as true but only one of them (which ever is selected first) are enforced at the db level.

Comment (mahendraHegde) [reply app/src/modules/settings/routes/data-model/field-detail/field-detail-advanced/field-detail-advanced-schema.vue]: > unique or index selected -> all drivers return only the respective field as true

in this case users might select the other checkbox, but always get error on unsupported driver. 

> unique + index are selected
driver supports both -> both returned as true and set at db level

we could do that, but that would be redundant as unique itself is a superset of simple index. 

imo, it makes sense to always return true for `index` if it is `unique`, but returning unique=true when its indexed would give wrong info. so we have to handle it at UI side or users will select it and either nothing happens(on backend i already ignore `index` request when field is already unique), or we throw an error .

**Comment by ComfortablyCoding** [comment]
> hey @ComfortablyCoding thanks for the review. I thought of using only 1 field instead object, however when any field is updated, UI sends the indexed as null, and we wont know if the intention is to remove the index or just a null, also its consistent with other fields(ex: unique), we wont remove the index unless ui sends false.

Fair point, in that case we should hold off on this PR until we fix this issue of values being sent even though they are not changed. This is not an expected behaviour from the app. Once done, we can then re-visit this and use a single field which is the ideal case here. 

> i didnt see any tests for the files i have made changes to, which tests are you pointing to?

Looks like this can be disregarded, we do not seem to have test coverage for this file.

> i will change it to INDEX_NAME to keep consistent with other drivers.

Let's hold off on this for now as this PR is paused. It will be removed anyway if the proposed changes are implemented.

**Comment by ComfortablyCoding** [comment]
Thanks for the contribution, I am closing this in favour of https://github.com/directus/directus/pull/23149. I have cherry picked any relevant code blocks with attribution.

**Pull request closed without merging**
